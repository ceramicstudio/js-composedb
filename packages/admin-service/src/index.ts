export { router } from './router.js'
export type { Router } from './router.js'
export { GraphQLQueryCodec, Service } from './service.js'
export type { GraphQLResult, ServiceConfig, ServiceParams } from './service.js'
