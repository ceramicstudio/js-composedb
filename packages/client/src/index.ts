/**
 * High-level ComposeDB client, based on the {@linkcode runtime ComposeDB runtime}.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/client
 * ```
 *
 * @module client
 */

export { ComposeClient, type ComposeClientParams } from './client.js'
