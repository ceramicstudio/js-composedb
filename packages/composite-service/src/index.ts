export { router } from './router.js'
export type { CreateModelInput, LoadModelInput, SaveCompositeInput, Router } from './router.js'
export { Service } from './service.js'
export type { GraphQLResult, SavedComposite, ServiceParams } from './service.js'
