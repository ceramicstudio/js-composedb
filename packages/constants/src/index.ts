export const NODE_ID_FIELD = 'id'

export const VIEWER_ID_HEADER = 'ComposeDB-ViewerID'
