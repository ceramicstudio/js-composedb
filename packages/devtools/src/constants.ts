export const DOC_ID_FIELD = 'id'
