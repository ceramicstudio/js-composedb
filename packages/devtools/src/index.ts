/**
 * Development tools library.
 *
 * ## Installation
 *
 * ```sh
 * npm install --dev @composedb/devtools
 * ```
 *
 * @module devtools
 */

export * from './composite.js'
export * from './formats/json.js'
export { createRuntimeDefinition, getName } from './formats/runtime.js'
export { createAbstractCompositeDefinition } from './schema/compiler.js'
export * from './schema/types.js'
export * from './test-utils.js'
export * from './types.js'
export * from './utils.js'
