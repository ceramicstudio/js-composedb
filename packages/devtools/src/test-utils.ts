import type { ModelDefinition, ModelViewsDefinition } from '@ceramicnetwork/stream-model'
import type { InternalCompositeDefinition } from '@composedb/types'

import { createAbstractCompositeDefinition } from './schema/compiler.js'
import type { AbstractCompositeDefinition } from './schema/types.js'

/** @internal */
export function mockDefinition(
  definition: AbstractCompositeDefinition,
  providedModels: Record<string, ModelDefinition> = {}
): InternalCompositeDefinition {
  const models: Record<string, ModelDefinition> = {}
  const modelsViews: Record<string, ModelViewsDefinition> = {}

  for (const abstractModel of Object.values(definition.models)) {
    if (abstractModel.action === 'create') {
      const definition = abstractModel.model
      models[`${definition.name}ID`] = definition
    } else {
      const definition = providedModels[abstractModel.id]
      if (definition == null) {
        throw new Error(`Missing provided model ${abstractModel.id}`)
      }
      models[abstractModel.id] = definition
      modelsViews[abstractModel.id] = abstractModel.views
    }
  }

  return {
    version: '1.0',
    commonEmbeds: definition.commonEmbeds,
    models,
    views: { models: modelsViews },
  }
}

/** @internal */
export function mockDefinitionFromSchema(
  schema: string,
  providedModels?: Record<string, ModelDefinition>
): InternalCompositeDefinition {
  return mockDefinition(createAbstractCompositeDefinition(schema), providedModels)
}
