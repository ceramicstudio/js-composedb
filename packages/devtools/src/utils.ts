import type { ModelRelationViewDefinition, ModelViewDefinition } from '@ceramicnetwork/stream-model'
import type { StreamRef } from '@ceramicnetwork/streamid'
import type { RuntimeRelation, RuntimeRelationSource, RuntimeViewField } from '@composedb/types'

/** @internal */
export function streamIDToString(id: StreamRef | string): string {
  return typeof id === 'string' ? id : id.toString()
}

/** @internal */
export function applyMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => R): Record<keyof M, R> {
  return Object.entries(inputs).reduce((acc, [key, value]) => {
    acc[key as keyof M] = callFunc(value as V)
    return acc
  }, {} as Record<keyof M, R>)
}

/** @internal */
export async function promiseMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => Promise<R>): Promise<Record<keyof M, R>> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  const results = await Promise.all(Object.values(inputs).map((value) => callFunc(value as any)))
  return Object.keys(inputs).reduce((acc, key, i) => {
    acc[key as keyof M] = results[i]
    return acc
  }, {} as Record<keyof M, R>)
}

type RelationViewType = ModelRelationViewDefinition['type']

const RELATION_VIEW_SOURCES: Record<RelationViewType, RuntimeRelationSource> = {
  relationDocument: 'document',
  relationFrom: 'queryConnection',
  relationCountFrom: 'queryCount',
}

const RELATION_VIEW_TYPES = Object.keys(RELATION_VIEW_SOURCES)

/** @internal */
export function assertRelationViewDefinition(
  view: ModelViewDefinition
): ModelRelationViewDefinition {
  if (!RELATION_VIEW_TYPES.includes(view.type)) {
    throw new Error(`Invalid relation view: ${view.type}`)
  }
  return view as ModelRelationViewDefinition
}

/** @internal */
export function viewDefinitionToRuntimeRelation(view: ModelViewDefinition): RuntimeRelation {
  const relation = assertRelationViewDefinition(view)
  return {
    source: RELATION_VIEW_SOURCES[relation.type],
    model: relation.model,
    property: relation.property,
  }
}

/** @internal */
export function viewDefinitionToRuntime(view: ModelViewDefinition): RuntimeViewField {
  switch (view.type) {
    case 'documentAccount':
    case 'documentVersion':
      return { type: 'view', viewType: view.type }
    case 'relationCountFrom':
    case 'relationDocument':
    case 'relationFrom':
      return {
        type: 'view',
        viewType: 'relation',
        relation: viewDefinitionToRuntimeRelation(view),
      }
    default:
      // @ts-ignore unexpected view type
      throw new Error(`Unsupported view type: ${view.type as string}`)
  }
}

/** @internal */
export function viewRuntimeToModel(field: RuntimeViewField): ModelViewDefinition {
  if (field.viewType === 'relation') {
    const { model, property, source } = field.relation
    switch (source) {
      case 'document':
        return { type: 'relationDocument', model, property }
      case 'queryConnection':
        return { type: 'relationFrom', model, property }
      case 'queryCount':
        return { type: 'relationCountFrom', model, property }
    }
  }
  return { type: field.viewType }
}
