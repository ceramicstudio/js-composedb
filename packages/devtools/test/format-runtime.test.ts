/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import {
  createCommentSchemaWithPost,
  loadPostSchemaWithComments,
  noteSchema,
  postSchema,
  profilesSchema,
} from '@composedb/test-schemas'

import { createRuntimeDefinition, getName, mockDefinitionFromSchema } from '../src'

describe('Runtime format', () => {
  const profilesDefinition = mockDefinitionFromSchema(profilesSchema)

  describe('getName()', () => {
    test('converts input to pascal case', () => {
      expect(getName('Foo bar')).toBe('FooBar')
      expect(getName('foo_bar')).toBe('FooBar')
      expect(getName('Foo-bar')).toBe('FooBar')
    })

    test('adds the prefix', () => {
      expect(getName('bar', 'Foo')).toBe('FooBar')
      expect(getName('foo_bar', 'Bar')).toBe('BarFooBar')
    })

    test('ignores the prefix if already included', () => {
      expect(getName('foo_bar', 'Foo')).toBe('FooBar')
    })
  })

  test('Profile - multiples models with common local references', () => {
    const runtime = createRuntimeDefinition(profilesDefinition)
    expect(runtime).toMatchSnapshot()
  })

  test('Profile definition with added model view', () => {
    const runtime = createRuntimeDefinition({
      ...profilesDefinition,
      views: { models: { GenericProfileID: { version: { type: 'documentVersion' } } } },
    })
    expect(runtime).toMatchSnapshot()
  })

  test('Note model definition with views', () => {
    const noteDefinition = mockDefinitionFromSchema(noteSchema)
    const runtime = createRuntimeDefinition(noteDefinition)
    expect(runtime).toMatchSnapshot()
  })

  test('Post with comments relations', () => {
    const postDefinition = mockDefinitionFromSchema(postSchema)
    expect(createRuntimeDefinition(postDefinition)).toMatchSnapshot()

    const commentDefinition = mockDefinitionFromSchema(
      createCommentSchemaWithPost('PostID'),
      postDefinition.models
    )
    expect(createRuntimeDefinition(commentDefinition)).toMatchSnapshot()

    const postWithCommentsDefinition = mockDefinitionFromSchema(
      loadPostSchemaWithComments('PostID', 'CommentID'),
      commentDefinition.models
    )
    expect(createRuntimeDefinition(postWithCommentsDefinition)).toMatchSnapshot()
  })
})
