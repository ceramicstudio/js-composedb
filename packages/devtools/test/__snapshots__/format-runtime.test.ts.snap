// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtime format Note model definition with views 1`] = `
{
  "accountData": {
    "noteList": {
      "name": "Note",
      "type": "connection",
    },
  },
  "enums": {
    "NoteStatus": [
      "DEFAULT",
      "IMPORTANT",
      "ARCHIVE",
    ],
  },
  "models": {
    "Note": {
      "accountRelation": {
        "type": "list",
      },
      "id": "NoteID",
    },
  },
  "objects": {
    "Note": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "status": {
        "refName": "NoteStatus",
        "refType": "enum",
        "required": false,
        "type": "reference",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "title": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
  },
}
`;

exports[`Runtime format Post with comments relations 1`] = `
{
  "accountData": {
    "postList": {
      "name": "Post",
      "type": "connection",
    },
  },
  "enums": {},
  "models": {
    "Post": {
      "accountRelation": {
        "type": "list",
      },
      "id": "PostID",
    },
  },
  "objects": {
    "Post": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "date": {
        "required": false,
        "type": "datetime",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "title": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
  },
}
`;

exports[`Runtime format Post with comments relations 2`] = `
{
  "accountData": {
    "commentList": {
      "name": "Comment",
      "type": "connection",
    },
    "postList": {
      "name": "Post",
      "type": "connection",
    },
  },
  "enums": {},
  "models": {
    "Comment": {
      "accountRelation": {
        "type": "list",
      },
      "id": "CommentID",
    },
    "Post": {
      "accountRelation": {
        "type": "list",
      },
      "id": "PostID",
    },
  },
  "objects": {
    "Comment": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "post": {
        "relation": {
          "model": "PostID",
          "property": "postID",
          "source": "document",
        },
        "type": "view",
        "viewType": "relation",
      },
      "postID": {
        "required": true,
        "type": "streamid",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
    "Post": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "date": {
        "required": false,
        "type": "datetime",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "title": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
  },
}
`;

exports[`Runtime format Post with comments relations 3`] = `
{
  "accountData": {
    "commentList": {
      "name": "Comment",
      "type": "connection",
    },
    "postList": {
      "name": "Post",
      "type": "connection",
    },
  },
  "enums": {},
  "models": {
    "Comment": {
      "accountRelation": {
        "type": "list",
      },
      "id": "CommentID",
    },
    "Post": {
      "accountRelation": {
        "type": "list",
      },
      "id": "PostID",
    },
  },
  "objects": {
    "Comment": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "post": {
        "relation": {
          "model": "PostID",
          "property": "postID",
          "source": "document",
        },
        "type": "view",
        "viewType": "relation",
      },
      "postID": {
        "required": true,
        "type": "streamid",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
    "Post": {
      "author": {
        "type": "view",
        "viewType": "documentAccount",
      },
      "comments": {
        "relation": {
          "model": "CommentID",
          "property": "postID",
          "source": "queryConnection",
        },
        "type": "view",
        "viewType": "relation",
      },
      "commentsCount": {
        "relation": {
          "model": "CommentID",
          "property": "postID",
          "source": "queryCount",
        },
        "type": "view",
        "viewType": "relation",
      },
      "date": {
        "required": false,
        "type": "datetime",
      },
      "text": {
        "required": true,
        "type": "string",
      },
      "title": {
        "required": true,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
  },
}
`;

exports[`Runtime format Profile - multiples models with common local references 1`] = `
{
  "accountData": {
    "genericProfile": {
      "name": "GenericProfile",
      "type": "node",
    },
    "personProfile": {
      "name": "PersonProfile",
      "type": "node",
    },
    "socialProfile": {
      "name": "SocialProfile",
      "type": "node",
    },
  },
  "enums": {},
  "models": {
    "GenericProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "GenericProfileID",
    },
    "PersonProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "PersonProfileID",
    },
    "SocialProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "SocialProfileID",
    },
  },
  "objects": {
    "GenericProfile": {
      "image": {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "name": {
        "required": false,
        "type": "string",
      },
    },
    "ImageMetadata": {
      "height": {
        "required": true,
        "type": "integer",
      },
      "mimeType": {
        "required": true,
        "type": "string",
      },
      "size": {
        "required": false,
        "type": "integer",
      },
      "src": {
        "required": true,
        "type": "string",
      },
      "width": {
        "required": true,
        "type": "integer",
      },
    },
    "ImageSources": {
      "alternatives": {
        "item": {
          "refName": "ImageMetadata",
          "refType": "object",
          "required": false,
          "type": "reference",
        },
        "required": false,
        "type": "list",
      },
      "original": {
        "refName": "ImageMetadata",
        "refType": "object",
        "required": true,
        "type": "reference",
      },
    },
    "PersonProfile": {
      "affiliations": {
        "item": {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "birthDate": {
        "required": false,
        "type": "date",
      },
      "gender": {
        "required": false,
        "type": "string",
      },
      "homeLocation": {
        "required": false,
        "type": "string",
      },
      "nationalities": {
        "item": {
          "required": false,
          "type": "countrycode",
        },
        "required": false,
        "type": "list",
      },
      "residenceCountry": {
        "required": false,
        "type": "countrycode",
      },
    },
    "SocialProfile": {
      "background": {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "description": {
        "required": false,
        "type": "string",
      },
      "emoji": {
        "required": false,
        "type": "string",
      },
      "url": {
        "required": false,
        "type": "string",
      },
    },
  },
}
`;

exports[`Runtime format Profile definition with added model view 1`] = `
{
  "accountData": {
    "genericProfile": {
      "name": "GenericProfile",
      "type": "node",
    },
    "personProfile": {
      "name": "PersonProfile",
      "type": "node",
    },
    "socialProfile": {
      "name": "SocialProfile",
      "type": "node",
    },
  },
  "enums": {},
  "models": {
    "GenericProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "GenericProfileID",
    },
    "PersonProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "PersonProfileID",
    },
    "SocialProfile": {
      "accountRelation": {
        "type": "single",
      },
      "id": "SocialProfileID",
    },
  },
  "objects": {
    "GenericProfile": {
      "image": {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "name": {
        "required": false,
        "type": "string",
      },
      "version": {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
    "ImageMetadata": {
      "height": {
        "required": true,
        "type": "integer",
      },
      "mimeType": {
        "required": true,
        "type": "string",
      },
      "size": {
        "required": false,
        "type": "integer",
      },
      "src": {
        "required": true,
        "type": "string",
      },
      "width": {
        "required": true,
        "type": "integer",
      },
    },
    "ImageSources": {
      "alternatives": {
        "item": {
          "refName": "ImageMetadata",
          "refType": "object",
          "required": false,
          "type": "reference",
        },
        "required": false,
        "type": "list",
      },
      "original": {
        "refName": "ImageMetadata",
        "refType": "object",
        "required": true,
        "type": "reference",
      },
    },
    "PersonProfile": {
      "affiliations": {
        "item": {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "birthDate": {
        "required": false,
        "type": "date",
      },
      "gender": {
        "required": false,
        "type": "string",
      },
      "homeLocation": {
        "required": false,
        "type": "string",
      },
      "nationalities": {
        "item": {
          "required": false,
          "type": "countrycode",
        },
        "required": false,
        "type": "list",
      },
      "residenceCountry": {
        "required": false,
        "type": "countrycode",
      },
    },
    "SocialProfile": {
      "background": {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "description": {
        "required": false,
        "type": "string",
      },
      "emoji": {
        "required": false,
        "type": "string",
      },
      "url": {
        "required": false,
        "type": "string",
      },
    },
  },
}
`;
